src/datanode/src/instance/flight/stream.rs

```diff
@@ -0,0 +1,184 @@
+// Copyright 2022 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+use arrow_flight::utils::flight_data_from_arrow_batch;
+use arrow_flight::{FlightData, SchemaAsIpc};
+use common_recordbatch::SendableRecordBatchStream;
+use common_telemetry::warn;
+use datatypes::arrow;
+use futures::channel::mpsc;
+use futures::channel::mpsc::Sender;
+use futures::{SinkExt, Stream, StreamExt};
+use pin_project::{pin_project, pinned_drop};
+use snafu::ResultExt;
+use tokio::task::JoinHandle;
+
+use crate::error;
+use crate::instance::flight::TonicResult;
+
+#[pin_project(PinnedDrop)]
+pub(super) struct GetStream {
```

Comment:
I think using async_stream's [`stream`](https://docs.rs/async-stream/0.1.2/async_stream/macro.stream.html) /[`try_stream`](https://docs.rs/async-stream/0.1.2/async_stream/macro.try_stream.html) wil also do the trick, for example:

```rust
fn create_tonic_stream_from_record_batch(
    mut recordbatches: SendableRecordBatchStream,
) -> Pin<Box<dyn Stream<Item = TonicResult<FlightData>>>> {
    let s = async_stream::stream! {
        while let Some(batch_or_err) = recordbatches.next().await {
            match batch_or_err {
                Ok(batch) => {
                    let (flight_dictionaries, flight_batch) =
                        flight_data_from_arrow_batch(batch.df_record_batch(), &arrow::ipc::writer::IpcWriteOptions::default());
                    debug_assert_eq!(flight_dictionaries.len(), 0);
                    yield Ok(flight_batch)
                }
                Err(e) => {
                    // handle the error
                    return;
                }
            }
        }
    };

    Box::pin(s)
}
```

Reply:
I think `recordbatches.next` should be evaluated in a separate runtime other than the GRPC runtime that drives Tonic. If use async_stream, it seems we can't achieve the goal? Polling the async_stream then would be done in the GRPC runtime.